### PR TITLE
Fix cross compiling

### LINKED ISSUE #46: Cross compilation failing with linker errors
I have been trying to compile some Rust software using a single-file Dockerfile and `docker buildx` targeting `x86_64` and `aarch64`. When building for the architecture that matches the host architecture, the build succeeds. But when the target architecture is different, it appears that libsasl still is compiling the objects using the host architecture, resulting in linking errors when the final link is attempted.

Based on my limited research and understanding, EM 62 means that the object file was in `x86_64` instead of `aarch64`.
```
#22 11.14   = note: /usr/lib/gcc-cross/aarch64-linux-gnu/10/../../../../aarch64-linux-gnu/bin/ld: /app/target/aarch64-unknown-linux-gnu/release/deps/libsasl2_sys-adc1f77f68e58ca4.rlib(client.o): Relocations in generic ELF (EM: 62)
#22 11.14           /usr/lib/gcc-cross/aarch64-linux-gnu/10/../../../../aarch64-linux-gnu/bin/ld: /app/target/aarch64-unknown-linux-gnu/release/deps/libsasl2_sys-adc1f77f68e58ca4.rlib(client.o): Relocations in generic ELF (EM: 62)
#22 11.14           /usr/lib/gcc-cross/aarch64-linux-gnu/10/../../../../aarch64-linux-gnu/bin/ld: /app/target/aarch64-unknown-linux-gnu/release/deps/libsasl2_sys-adc1f77f68e58ca4.rlib: error adding symbols: file in wrong format
#22 11.14           collect2: error: ld returned 1 exit status
#22 11.14
#22 11.14
#22 11.18 error: could not compile `my-app` due to previous error
```

Following the examples in the `cross` dockerfiles, I do have `CC_aarch64_unknown_linux_gnu` set to `aarch64-linux-gnu-gcc`. I did try adding `--build=${HOST}` into the `build.rs` script, but that doesn't seem to have had an impact, though `configure` then does report that it should cross compile (without `--build`, `configure` reports that it is not cross compiling.)

I was able to get a good compilation by setting `CC` in addition to `CC_aarch64_unknown_linux_gnu`. I'm wondering if there is some reason that the configure script isn't picking up that it should be compiling for the target rather than the current host. Perhaps `CC` _is_ the way I should do this, but I wanted to confirm; I'm not well-versed in cross compilation.

`Dockerfile`:
```dockerfile
# syntax=docker/dockerfile:1

FROM --platform=$BUILDPLATFORM rust:1.61 as build
ARG TARGETARCH
ARG BUILDARCH
COPY platform.sh .
RUN ./platform.sh

RUN dpkg --add-architecture $TARGETARCH && \
    apt-get update && \
    apt-get install --assume-yes --no-install-recommends $(cat /.deps)

# create a new empty shell project
RUN cargo new --bin /app
WORKDIR /app

# copy over your manifests
COPY ./Cargo.lock ./Cargo.lock
COPY ./Cargo.toml ./Cargo.toml

# this build step will cache your dependencies
RUN . /.env && \
    cargo build --locked --release --target $(cat /.platform) -vv

# copy your source tree
COPY ./src ./src

# build for release
RUN . /.env && \
    cargo build --locked --release --target $(cat /.platform) --verbose && \
    mkdir bin && \
    cp target/$(cat /.platform)/release/my-app bin/my-app

# The final base image
FROM gcr.io/distroless/cc
COPY --from=build /app/bin/my-app /
CMD ["/my-app"]
```

`platform.sh`:
```bash
#!/bin/bash

# When targeting x86_64, we target the Haswell CPU architecture as a least
# common denominator among the supported CPU architectures on AWS. This way
# we can get some reasonable CPU-related performance benefits.

set -euo pipefail

case $BUILDARCH in
    "amd64")
        cat <<EOF > $CARGO_HOME/config.toml
[target.x86_64-unknown-linux-gnu]
rustflags = "-C target-cpu=haswell"

[target.aarch64-unknown-linux-gnu]
rustflags = "-C target-cpu=neoverse-n1 -C target-feature=+lse"
linker = "aarch64-linux-gnu-gcc"
EOF
        ;;
    "arm64")
        cat <<EOF > $CARGO_HOME/config.toml
[target.x86_64-unknown-linux-gnu]
rustflags = "-C target-cpu=haswell"
linker = "x86_64-linux-gnu-gcc"

[target.aarch64-unknown-linux-gnu]
rustflags = "-C target-cpu=neoverse-n1 -C target-feature=+lse"
EOF
        ;;
esac

case $TARGETARCH in
    "amd64")
        echo "x86_64-unknown-linux-gnu" > /.platform
        ;;
    "arm64")
        echo "aarch64-unknown-linux-gnu" > /.platform
        ;;
esac

touch /.deps
case "$BUILDARCH:$TARGETARCH" in
    "arm64:amd64")
        echo "crossbuild-essential-amd64 g++-x86-64-linux-gnu libc6-dev-amd64-cross autoconf automake binutils ca-certificates curl file gcc git libtool m4 make g++ libc6-dev libclang-dev pkg-config cmake libssl-dev:$TARGETARCH" > /.deps
        cat <<'EOF' > /.env
export CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU_LINKER=x86_64-linux-gnu-gcc
export CC_x86_64_unknown_linux_gnu=x86_64-linux-gnu-gcc
#export CC=x86_64-linux-gnu-gcc
export CXX_x86_64_unknown_linux_gnu=x86_64-linux-gnu-g++
#export CXX=x86_64-linux-gnu-g++
export BINDGEN_EXTRA_CLANG_ARGS_x86_64_unknown_linux_gnu="--sysroot=/usr/x86_64-linux-gnu"
export PKG_CONFIG_PATH="/usr/lib/x86_64-linux-gnu/pkgconfig${PKG_CONFIG_PATH:+:}${PKG_CONFIG_PATH:-}"
export PKG_CONFIG_ALLOW_CROSS=1
export OPENSSL_INCLUDE_DIR=/usr/include/openssl
export X86_64_UNKNOWN_LINUX_GNU_OPENSSL_LIB_DIR=/usr/lib/aarch64-linux-gnu
EOF
        ;;
    "amd64:arm64")
        echo "crossbuild-essential-arm64 g++-aarch64-linux-gnu libc6-dev-arm64-cross autoconf automake binutils ca-certificates curl file gcc git libtool m4 make g++ libc6-dev libclang-dev pkg-config cmake libssl-dev:$TARGETARCH" > /.deps
        cat <<'EOF' > /.env
export CARGO_TARGET_AARCH64_UNKNOWN_LINUX_GNU_LINKER=aarch64-linux-gnu-gcc
export CC_aarch64_unknown_linux_gnu=aarch64-linux-gnu-gcc
#export CC=aarch64-linux-gnu-gcc
export CXX_aarch64_unknown_linux_gnu=aarch64-linux-gnu-g++
#export CXX=aarch64-linux-gnu-g++
export BINDGEN_EXTRA_CLANG_ARGS_aarch64_unknown_linux_gnu="--sysroot=/usr/aarch64-linux-gnu"
export PKG_CONFIG_PATH="/usr/lib/aarch64-linux-gnu/pkgconfig${PKG_CONFIG_PATH:+:}${PKG_CONFIG_PATH:-}"
export PKG_CONFIG_ALLOW_CROSS=1
export OPENSSL_INCLUDE_DIR=/usr/include/openssl
export AARCH64_UNKNOWN_LINUX_GNU_OPENSSL_LIB_DIR=/usr/lib/aarch64-linux-gnu
EOF
        ;;
    *)
        echo "autoconf automake binutils ca-certificates curl file gcc git libtool m4 make g++ libc6-dev libclang-dev pkg-config cmake libssl-dev:$TARGETARCH" > /.deps
        cat <<EOF > /.env
EOF
        ;;
esac

rustup target add $(cat /.platform)

echo "Cargo Configuration:"
cat $CARGO_HOME/config.toml
echo "Target triple: $(cat /.platform)"
echo "Dependencies to install: $(cat /.deps)"
```

### PR DESCRIPTION
**Pull request by simlay**
Closes #46.

### CODE REVIEW

File: sasl2-sys/build.rs [hunk 1]
```diff
@@ -113,15 +115,18 @@ fn build_sasl(metadata: &Metadata) {
...
     if metadata.target.contains("darwin") {
         configure_args.push("--disable-macos-framework".into());
     }
     if metadata.host != metadata.target {
+        configure_args.push(format!("--build={}", metadata.host));
         configure_args.push(format!("--host={}", metadata.target));
     }
     cmd(src_dir.join("configure"), &configure_args)
         .dir(&src_dir)
+        .env("ac_cv_gssapi_supports_spnego", "yes")
```
Comment (simlay) [inline sasl2-sys/build.rs:129]: This should actually be a argument when `metadata.host != metadata.target`.

**Comment by jszwedko** [comment]
We just ran into this again today on [vector](https://github.com/vectordotdev/vector), https://github.com/vectordotdev/vector/issues/16737#issuecomment-1462027042. Is there any chance this could be picked up and merged @benesch ?

**Comment by jszwedko** [comment]
Apologies for the nuisance, but just wondering what you think of this change @benesch . We are still blocked from supporting GSSAPI authentication for Kafka on non-x86_64 platforms.

**Review by benesch — COMMENTED** [review]
Apologies, I'm afraid I won't have time to review this anytime soon. These build scripts are so finicky so it's hard for me to merge this without taking a fine-tooth comb to it. I'm afraid of breaking things for folks who aren't cross-compiling. I've left a few comments within, but I'm afraid your best bet if you need this ASAP is to use a patch directive to override the sasl2-sys dependency to a Git repository under your control.

I'll leave a few comments within now though. One question I have is: can you make this work by setting `CC` and `ac_cv_gssapi_supports_spnego=yes` in your environment before invoking `cargo build`? That's how we've made cross-compiling work at Materialize with this package. You can see our code for doing so here: https://github.com/MaterializeInc/materialize/blob/v0.26.6/misc/python/materialize/xcompile.py#L52-L59

File: sasl2-sys/build.rs [hunk 2]
```diff
@@ -113,17 +115,20 @@ fn build_sasl(metadata: &Metadata) {
         "--with-pic".into(),
         format!("CPPFLAGS={}", cppflags),
         format!("CFLAGS={}", cflags),
+        format!("CC={}", compiler_path.to_str().unwrap_or_default()),
     ];
     if metadata.target.contains("darwin") {
         configure_args.push("--disable-macos-framework".into());
     }
     if metadata.host != metadata.target {
+        configure_args.push(format!("--build={}", metadata.host));
```
Comment (benesch) [inline sasl2-sys/build.rs:124]: Why does this need to be specified? Doesn't `--build` always default to the platform on which you're running `configure`?

**Comment by benesch** [comment]
Closing due to staleness.

**Pull request closed without merging**